File: apps/video/src/youtube.test.ts
import {
  extractYoutubeUrls,
  getYoutubeTiming,
  parseYoutubeTimestamp,
  youtubeVideoId,
} from "@coldbrew/packages/youtube.js";
import { afterEach, describe, expect, it, vi } from "vitest";

afterEach(() => vi.unstubAllGlobals());

describe("findYoutubeUrls", () => {
  it("returns unique YouTube URLs from a donation message", () => {
    const result = extractYoutubeUrls(
      "Play https://youtu.be/dQw4w9WgXcQ! Also www.youtube.com/watch?v=dQw4w9WgXcQ and https://youtu.be/dQw4w9WgXcQ",
    );
    expect(result).toEqual([
      "https://youtu.be/dQw4w9WgXcQ",
      "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
    ]);
  });

  it("ignores non-YouTube URLs and empty messages", () => {
    expect(extractYoutubeUrls("https://notyoutube.com/watch?v=123")).toEqual([]);
    expect(extractYoutubeUrls(null)).toEqual([]);
  });

  it("replaces http with https", () => {
    const result = extractYoutubeUrls("http://youtu.be/dQw4w9WgXcQ");
    expect(result).toEqual(["https://youtu.be/dQw4w9WgXcQ"]);
  });
});

describe("youtubeVideoId", () => {
  it("extracts IDs from supported YouTube URLs", () => {
    expect(youtubeVideoId("https://youtu.be/dQw4w9WgXcQ")).toBe("dQw4w9WgXcQ");
    expect(youtubeVideoId("https://www.youtube.com/shorts/dQw4w9WgXcQ")).toBe("dQw4w9WgXcQ");
  });

  it("rejects invalid and non-YouTube URLs", () => {
    expect(youtubeVideoId("not a url")).toBeNull();
    expect(youtubeVideoId("https://example.com/watch?v=dQw4w9WgXcQ")).toBeNull();
  });
});

describe("YouTube timing", () => {
  it.each([
    ["90", 90],
    ["1m30s", 90],
    ["2h3m4s", 7384],
    ["15s", 15],
  ])("parses %s", (value, expected) => {
    expect(parseYoutubeTimestamp(value)).toBe(expected);
  });

  it.each([null, "", "1:30", "abc", "-1", "1m30"])("rejects %s", (value) => {
    expect(parseYoutubeTimestamp(value)).toBeNull();
  });

  it("uses the exact video length", async () => {
    vi.stubGlobal("fetch", vi.fn().mockResolvedValue(new Response('{"lengthSeconds":"213"}')));

    const $timing = getYoutubeTiming("https://youtu.be/dQw4w9WgXcQ");

    await expect($timing).resolves.toMatchObject({
      value: { startSeconds: 0, endSeconds: 213, durationSeconds: 213 },
    });
  });

  it("falls back to the stream duration when video details are omitted", async () => {
    vi.stubGlobal(
      "fetch",
      vi.fn().mockResolvedValue(new Response('{"approxDurationMs":"7260183"}')),
    );

    const $timing = getYoutubeTiming("https://www.youtube.com/watch?v=_JXL6Fn99l8&t=13s");

    await expect($timing).resolves.toMatchObject({
      value: { startSeconds: 0, endSeconds: 7260, durationSeconds: 7260 },
    });
  });

  it("falls back to player metadata when the page omits all duration fields", async () => {
    const fetchMock = vi
      .fn()
      .mockResolvedValueOnce(new Response('{"INNERTUBE_CLIENT_VERSION":"2.20260824.01.00"}'))
      .mockResolvedValueOnce(new Response('{"videoDetails":{"lengthSeconds":"7260"}}'));
    vi.stubGlobal("fetch", fetchMock);

    const $timing = getYoutubeTiming("https://www.youtube.com/watch?v=_JXL6Fn99l8&t=13s");

    await expect($timing).resolves.toMatchObject({
      value: { startSeconds: 0, endSeconds: 7260, durationSeconds: 7260 },
    });
    expect(fetchMock).toHaveBeenCalledTimes(2);
    expect(fetchMock).toHaveBeenLastCalledWith(
      "https://www.youtube.com/youtubei/v1/player?prettyPrint=false",
      expect.objectContaining({ method: "POST" }),
    );
  });

  it("always starts at zero and uses a valid end parameter", async () => {
    vi.stubGlobal("fetch", vi.fn().mockResolvedValue(new Response('{"lengthSeconds":"213"}')));

    const $timing = getYoutubeTiming("https://www.youtube.com/watch?v=dQw4w9WgXcQ&t=1m15s&end=180");

    await expect($timing).resolves.toMatchObject({
      value: { startSeconds: 0, endSeconds: 180, durationSeconds: 213 },
    });
  });

  it("uses an explicitly requested range instead of URL timestamps", async () => {
    vi.stubGlobal("fetch", vi.fn().mockResolvedValue(new Response('{"lengthSeconds":"213"}')));

    const $timing = getYoutubeTiming("https://www.youtube.com/watch?v=dQw4w9WgXcQ&t=15s&end=180", {
      startSeconds: 30,
      endSeconds: 90,
    });

    await expect($timing).resolves.toMatchObject({
      value: { startSeconds: 30, endSeconds: 90, durationSeconds: 213 },
    });
  });

  it("uses the video length when the requested end is omitted", async () => {
    vi.stubGlobal("fetch", vi.fn().mockResolvedValue(new Response('{"lengthSeconds":"213"}')));

    const $timing = getYoutubeTiming("https://youtu.be/dQw4w9WgXcQ?end=180", {
      startSeconds: 30,
      endSeconds: null,
    });

    await expect($timing).resolves.toMatchObject({
      value: { startSeconds: 30, endSeconds: 213, durationSeconds: 213 },
    });
  });

  it.each([
    { startSeconds: 213, endSeconds: null },
    { startSeconds: 30, endSeconds: 30 },
    { startSeconds: 30, endSeconds: 214 },
  ])("rejects an invalid requested range: %o", async (requestedTiming) => {
    vi.stubGlobal("fetch", vi.fn().mockResolvedValue(new Response('{"lengthSeconds":"213"}')));

    const $timing = getYoutubeTiming("https://youtu.be/dQw4w9WgXcQ", requestedTiming);

    await expect($timing).resolves.toMatchObject({
      error: { type: "youtube: invalid timing" },
    });
  });

  it.each([
    ["https://youtu.be/dQw4w9WgXcQ?t=300", { startSeconds: 0, endSeconds: 213 }],
    ["https://youtu.be/dQw4w9WgXcQ?t=30&end=20", { startSeconds: 0, endSeconds: 20 }],
    ["https://youtu.be/dQw4w9WgXcQ?start=nope&end=300", { startSeconds: 0, endSeconds: 213 }],
  ])("falls back for invalid bounds in %s", async (url, expected) => {
    vi.stubGlobal("fetch", vi.fn().mockResolvedValue(new Response('{"lengthSeconds":"213"}')));

    const $timing = getYoutubeTiming(url);

    await expect($timing).resolves.toMatchObject({ value: expected });
  });

  it("rejects unsupported URLs without fetching them", async () => {
    const fetchMock = vi.fn();
    vi.stubGlobal("fetch", fetchMock);

    const $timing = getYoutubeTiming("https://example.com/watch?v=dQw4w9WgXcQ");

    await expect($timing).resolves.toMatchObject({ error: { type: "youtube: invalid url" } });
    expect(fetchMock).not.toHaveBeenCalled();
  });
});

File: internal/youtube/youtube_test.go
package youtube

import (
	"context"
	"io"
	"net/http"
	"reflect"
	"strings"
	"testing"
)

type roundTripFunc func(*http.Request) (*http.Response, error)

func (function roundTripFunc) RoundTrip(request *http.Request) (*http.Response, error) {
	return function(request)
}

func TestExtractURLs(t *testing.T) {
	actual := ExtractURLs("Play https://youtu.be/dQw4w9WgXcQ! Also www.youtube.com/watch?v=dQw4w9WgXcQ and https://youtu.be/dQw4w9WgXcQ")
	expected := []string{"https://youtu.be/dQw4w9WgXcQ", "https://www.youtube.com/watch?v=dQw4w9WgXcQ"}
	if !reflect.DeepEqual(actual, expected) {
		t.Fatalf("ExtractURLs() = %#v; want %#v", actual, expected)
	}
	if actual := ExtractURLs("https://notyoutube.com/watch?v=123"); len(actual) != 0 {
		t.Fatalf("expected non-YouTube URL to be ignored, got %#v", actual)
	}
	if actual := ExtractURLs("http://youtu.be/dQw4w9WgXcQ"); !reflect.DeepEqual(actual, []string{"https://youtu.be/dQw4w9WgXcQ"}) {
		t.Fatalf("expected HTTP URL to be upgraded, got %#v", actual)
	}
}

func TestVideoID(t *testing.T) {
	for rawURL, expected := range map[string]string{
		"https://youtu.be/dQw4w9WgXcQ":               "dQw4w9WgXcQ",
		"https://www.youtube.com/shorts/dQw4w9WgXcQ": "dQw4w9WgXcQ",
	} {
		actual, ok := VideoID(rawURL)
		if !ok || actual != expected {
			t.Fatalf("VideoID(%q) = %q, %v; want %q, true", rawURL, actual, ok, expected)
		}
	}
	for _, rawURL := range []string{"not a url", "https://example.com/watch?v=dQw4w9WgXcQ"} {
		if _, ok := VideoID(rawURL); ok {
			t.Fatalf("expected VideoID(%q) to reject URL", rawURL)
		}
	}
}

func TestParseTimestamp(t *testing.T) {
	valid := map[string]int{"90": 90, "1m30s": 90, "2h3m4s": 7384, "15s": 15}
	for value, expected := range valid {
		actual, ok := ParseTimestamp(value)
		if !ok || actual != expected {
			t.Fatalf("ParseTimestamp(%q) = %d, %v; want %d, true", value, actual, ok, expected)
		}
	}
	for _, value := range []string{"", "1:30", "abc", "-1", "1m30"} {
		if _, ok := ParseTimestamp(value); ok {
			t.Fatalf("expected ParseTimestamp(%q) to reject value", value)
		}
	}
}

func TestGetTiming(t *testing.T) {
	tests := []struct {
		name      string
		rawURL    string
		body      string
		requested *RequestedTiming
		expected  Timing
	}{
		{name: "exact length", rawURL: "https://youtu.be/dQw4w9WgXcQ", body: `{"lengthSeconds":"213"}`, expected: Timing{0, 213, 213}},
		{name: "approximate duration", rawURL: "https://www.youtube.com/watch?v=_JXL6Fn99l8&t=13s", body: `{"approxDurationMs":"7260183"}`, expected: Timing{0, 7260, 7260}},
		{name: "valid end", rawURL: "https://www.youtube.com/watch?v=dQw4w9WgXcQ&t=1m15s&end=180", body: `{"lengthSeconds":"213"}`, expected: Timing{0, 180, 213}},
		{name: "requested range", rawURL: "https://youtu.be/dQw4w9WgXcQ", body: `{"lengthSeconds":"213"}`, requested: &RequestedTiming{StartSeconds: 30, EndSeconds: integerPointer(90)}, expected: Timing{30, 90, 213}},
		{name: "requested open end", rawURL: "https://youtu.be/dQw4w9WgXcQ?end=180", body: `{"lengthSeconds":"213"}`, requested: &RequestedTiming{StartSeconds: 30}, expected: Timing{30, 213, 213}},
		{name: "invalid start falls back", rawURL: "https://youtu.be/dQw4w9WgXcQ?t=300", body: `{"lengthSeconds":"213"}`, expected: Timing{0, 213, 213}},
		{name: "invalid range uses end", rawURL: "https://youtu.be/dQw4w9WgXcQ?t=30&end=20", body: `{"lengthSeconds":"213"}`, expected: Timing{0, 20, 213}},
		{name: "invalid end falls back", rawURL: "https://youtu.be/dQw4w9WgXcQ?start=nope&end=300", body: `{"lengthSeconds":"213"}`, expected: Timing{0, 213, 213}},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			client := responseClient(test.body)
			actual, err := GetTiming(context.Background(), client, test.rawURL, test.requested)
			if err != nil {
				t.Fatal(err)
			}
			if actual != test.expected {
				t.Fatalf("GetTiming() = %#v; want %#v", actual, test.expected)
			}
		})
	}
}

func TestGetTimingFallsBackToPlayerMetadata(t *testing.T) {
	calls := 0
	client := &http.Client{Transport: roundTripFunc(func(request *http.Request) (*http.Response, error) {
		calls++
		body := `{"INNERTUBE_CLIENT_VERSION":"2.20260824.01.00"}`
		if calls == 2 {
			body = `{"videoDetails":{"lengthSeconds":"7260"}}`
			if request.Method != http.MethodPost || request.URL.String() != "https://www.youtube.com/youtubei/v1/player?prettyPrint=false" {
				t.Fatalf("unexpected player request: %s %s", request.Method, request.URL)
			}
		}
		return response(http.StatusOK, body), nil
	})}
	actual, err := GetTiming(context.Background(), client, "https://www.youtube.com/watch?v=_JXL6Fn99l8&t=13s", nil)
	if err != nil {
		t.Fatal(err)
	}
	if actual != (Timing{0, 7260, 7260}) || calls != 2 {
		t.Fatalf("GetTiming() = %#v after %d calls", actual, calls)
	}
}

func TestGetTimingRejectsInvalidRequestedRanges(t *testing.T) {
	for _, requested := range []RequestedTiming{
		{StartSeconds: 213},
		{StartSeconds: 30, EndSeconds: integerPointer(30)},
		{StartSeconds: 30, EndSeconds: integerPointer(214)},
	} {
		_, err := GetTiming(context.Background(), responseClient(`{"lengthSeconds":"213"}`), "https://youtu.be/dQw4w9WgXcQ", &requested)
		if err == nil || !strings.Contains(err.Error(), "youtube: invalid timing") {
			t.Fatalf("expected invalid timing error, got %v", err)
		}
	}
}

func TestGetTimingRejectsUnsupportedURLWithoutFetching(t *testing.T) {
	called := false
	client := &http.Client{Transport: roundTripFunc(func(*http.Request) (*http.Response, error) {
		called = true
		return response(http.StatusOK, ""), nil
	})}
	_, err := GetTiming(context.Background(), client, "https://example.com/watch?v=dQw4w9WgXcQ", nil)
	if err == nil || err.Error() != "youtube: invalid url" || called {
		t.Fatalf("expected invalid URL without fetch, got err=%v called=%v", err, called)
	}
}

func responseClient(body string) *http.Client {
	return &http.Client{Transport: roundTripFunc(func(*http.Request) (*http.Response, error) {
		return response(http.StatusOK, body), nil
	})}
}

func response(status int, body string) *http.Response {
	return &http.Response{StatusCode: status, Body: io.NopCloser(strings.NewReader(body)), Header: make(http.Header)}
}

func integerPointer(value int) *int { return &value }
